File: Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int> &height) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        stack<int> s;
        int i = 0, ret = 0;
        height.push_back(0);
        
        while (i < height.size()) {
            if (s.empty() || height[s.top()] <= height[i]) {
                s.push(i++);       
            } else {
                int k = s.top();
                s.pop();
                ret = max(ret, height[k] * (s.empty() ? i : i - 1 - s.top()));
                
            }
        }
        return ret;
    }
};

File: Largest Rectangle in Histogram.cs
public class Solution {
    public int LargestRectangleArea(int[] heights) {
        var stack = new Stack<int>(); // store index
        int result = 0;
        
        var i = 0;
        while (i < heights.Length || stack.Count > 0) {
            if (stack.Count == 0 || i < heights.Length && heights[stack.Peek()] <= heights[i]) {
                // the value of bottom index is minimum when scanning the heights.
                stack.Push(i++);
            } else {
                var j = stack.Pop();  // the minimum value's index is poped  at last.
                result = Math.Max(result, heights[j] * (stack.Count == 0 ? i : i - 1 - stack.Peek()));
            }
        }
        
        return result;
    }
}
